R/utilities-data-combined.R

```diff
@@ -1,4 +1,344 @@
+    return(NULL)
+  }
+
+  # Getting all datasets to have the same units.
+  combinedData <- .unitConverter(combinedData, xUnit, yUnit)
+
+  # Create observed versus simulated paired data using interpolation for each
+  # grouping level and combine the resulting data frames in a row-wise manner.
+  #
+  # Both of these routines will be carried out by `dplyr::group_modify()`.
+  pairedData <- combinedData %>%
+    dplyr::group_by(group) %>%
+    dplyr::group_modify(.f = ~ .extractResidualsToTibble(.x, scaling)) %>%
+    dplyr::ungroup()
+
+  return(pairedData)
+}
+
+#' Created observed versus simulated paired data
+#'
+#' @param data A data frame from `DataCombined$toDataFrame()`, which has been
+#'   further tidied using `.removeUnpairableDatasets()` and then
+#'   `.unitConverter()` functions.
+#'
+#' @examples
+#' # create an example data frame
+#' df <- dplyr::tibble(
+#'   dataType = c(rep("observed", 5), rep("simulated", 3)),
+#'   xValues = c(1, 3, 3.5, 4, 5, 0, 2, 4),
+#'   xUnit = ospUnits$Time$min,
+#'   xDimension = ospDimensions$Time,
+#'   yValues = c(1.9, 6.1, 7, 8.2, 1, 0, 4, 8),
+#'   yErrorValues = rnorm(8),
+#'   yUnit = ospUnits$`Concentration [mass]`$`mg/l`,
+#'   yDimension = ospDimensions$`Concentration (mass)`
+#' )
+#'
+#' ospsuite:::.extractResidualsToTibble(df)
+#'
+#' @keywords internal
+.extractResidualsToTibble <- function(data, scaling = tlf::Scaling$lin) {
+  # Since the data frames will be fed to `matrix()`, make sure that data has
+  # `data.frame` class. That is, if tibbles are supplied, coerce them to a
+  # simple data frame.
+  observedData <- as.data.frame(dplyr::filter(data, dataType == "observed"))
+  simulatedData <- as.data.frame(dplyr::filter(data, dataType == "simulated"))
+
+  # If available, error values will be useful for plotting error bars in the
+  # scatter plot. Even if not available, add missing values to be consistent.
+  if ("yErrorValues" %in% colnames(data)) {
+    yErrorValues <- data$yErrorValues[data$dataType == "observed"]
+  } else {
+    yErrorValues <- rep(NA_real_, nrow(observedData))
+  }
+
+  # Time matrix to match observed time with closest simulation time
+  # This method assumes that there simulated data are dense enough to capture observed data.
+  obsTimeMatrix <- matrix(observedData[, "xValues"], nrow(simulatedData), nrow(observedData), byrow = TRUE)
+  simTimeMatrix <- matrix(simulatedData[, "xValues"], nrow(simulatedData), nrow(observedData))
+
+  timeMatchedData <- as.numeric(sapply(as.data.frame(abs(obsTimeMatrix - simTimeMatrix)), which.min))
+
+  pairedData <- dplyr::tibble(
+    "xValues"      = observedData[, "xValues"],
+    "xUnit"        = unique(data$xUnit),
+    "xDimension"   = unique(data$xDimension),
+    "yValues"      = observedData[, "yValues"],
+    "yErrorValues" = yErrorValues,
+    "yUnit"        = unique(data$yUnit),
+    "yDimension"   = unique(data$yDimension),
+    "predValue"    = simulatedData[timeMatchedData, "yValues"]
+  )
```

Comment:
I wonder if it wouldn't be better for the understanding to rename the columns:
* X-Columns to Time-columns (e.g. `xValues`=>`timeValues `etc.) 
* Y-Columns to Observed-Columns (e.g. `yValues`=>`observedValues`)
* `predValue `=> `predictedValues `(in any case it  should be plural)

Reply:
I am just trying to be consistent with the naming schema present in the original data frame from which this paired data frame is extracted:

https://github.com/Open-Systems-Pharmacology/OSPSuite-R/blob/af89b3a242649124e077458ce29f37dd87516b72/R/data-combined.R#L378-L381

Yeah, I agree. Following names, although verbose, will be better:

- predValue -> predictedValues
- resValue -> residualValues

Reply:
Done.